### ISSUE TITLE
Wrong behaviour when setting / getting opacity on layers

### ISSUE THREAD

**Issue by nikitul** [mapgears/ol3-google-maps#164]
I'd like to thank you for this integration library.
Now the issue. 
I didn't try it on another properties but there is a chance to have the same behavior.

If I try to change layers Opacity property using layer.setOpacity( some_value) function, the value is applied on targeted layer BUT :
- the layer.getOpacity() function returns 0 and not the right value ; 
- after that ,  Opacity is changed again (i don't know the new value, but the listener on 'change:opacity' pops up twice) 

I've maid a jsFiddle with this behavior [here](http://jsfiddle.net/wn7dazqa/2/)

Thank You

**Comment by adube** [comment]
Thanks for your interests in ol3-google-maps.

Here's an explanation that causes the behaviour you're describing.  When olgm is activated, all layers in the ol map have their opacity automatically reduced to 0.  Each layer is then created as a `google.maps.OverlayView` and added directly to the Google map.  The original opacity is applied to that overlay.  Deactivating olgm would restore the original ol layer opacity.

The heralds are responsible of listening changes that occur in the original layer and apply them to the overlay.  See an example here: https://github.com/mapgears/ol3-google-maps/blob/master/src/herald/tilesourceherald.js#L268-L282

So, what you're describing may well be what's expected, because the original layer should stay at opacity `0` while olgm is active.  The overlay is the one that should get the opacity change.

Please, let us know if that makes sense.

**Comment by nikitul** [comment]
Yes, it makes sense :)

I ask you too much to guide me where I can read more about how to get googleTileLayer from ol3-google-maps based on ol3 layer? Is there a standard function or it should be done like in olgm.herald.TileSource.prototype.unwatchLayer ?

Thank you

**Comment by adube** [comment]
@nikitul I'm not sure I understand the question.  Would you please clarify?

**Comment by nikitul** [comment]
I'd like to use googleTileLayer and to set (or get) the opacity value with googleTileLayer.setOpacity().

For doing that, I need a function with ol3-layer as argument to get googleTileLayer from cache. 

Something similar i see in olgm.herald.TileSource.prototype.unwatchLayer function, but instead of unwatching the ol3-layer, I need to return googleTileLayer, (or the whole cacheItem). 
I did not see a function like that in olgm.herald.TileSource. Is there any?
Or is there an "official" way?

I am not very familiar with ol3-google-maps, and I don't know how to use it very well. 

Basically, what am I asking for is a "best practice" example.

Thank You again

**Comment by adube** [comment]
Thanks for the explanations.  I am not entirely sure if I understood everything correctly, though.  Sorry about that.  I'll try to give more explanations.

1) You say you want to have access to the google layer to be able to set its opacity.  In your example, are you talking about the google layer that renders the OSM map?  If so, then setting the opacity already works, no?  In that case, there would be no need to have a direct access to the google layer.

2) There's is currently no way through the OLGM code to have a direct access to the google layer.  If we wanted to be able to have access, I think the best way would be through the original OpenLayers layer.  For example, in your jsfiddle you define a `ol.layer.Tile` layer for OSM.  We could either add a custom property within the layer that would be a reference to the google layer, i.e. `mapMain.get('googleLayer')` for example.  That would be a solution A.  An other solution B could be to have a utility method that would receive the ol layer as argument, i.e. `olGM.getGoogleLayer(mapMain)` which would do the work of finding the google layer.  What do you think?

3) Finally, if you're talking about the changing the opacity of the actual base layer, i.e. the `olgm.layer.Google`, then that also is not currently supported natively.  Development would be required.

Please, let us what it is you would need.  If you wish to make a contribution, a PR would be welcome.  See also: https://github.com/mapgears/ol3-google-maps#our-commitment

**Comment by nikitul** [comment]
Thank you very much for your time.
I think that my English is not so elevate and I am sorry for that. But...
You understood me perfect.

Actually, what I want is to make a plugin for `ol.layer.Tile`, something like ol3-layerswitcher with the possibility of changing layers position using drag and drop, and also changing the layers opacity. So:

1. You have right when is about setting the opacity using `ol.layer.Tile`, everything it works. But If I want to get the value of opacity property (I use an input slider and I need that value for initialization), always I will get value 0 which is wrong. So I think it is a good idea to provide a way for doing that or at least to provide a way to inform the developer about the presence of ol3-google-maps. And here It comes what you explained me at point 2.

2. I think that both Ideas are good, because if I want to work with `ol.layer.Tile` (like in my case with the plugin I intended to do - and use ol3-google-maps just for info), I will use `mapMain.get('googleLayer')` to see if ol3-google-maps is present and get the opacity value from google layer, but if the plugin is intended to be used to work direct with ol3-google-maps, It is a great idea to have a `olGM.getGoogleLayer(mapMain)` function to work with.

3. I think that if it will be provided a function to get `olgm.layer.Google` is enough for the moment because if the developer have access to base layer, is easy to make his stuff work.

For the moment I resolved my problem by saving the opacity value as a property in my `mapMain` variable, and get the value from there with  `mapMain.get('myPlugin').opacity`.

**Comment by adube** [comment]
Cool!  I now I understand perfectly what and why you need what you need.  I think it would make sense to create a new issue about this feature in which a short explanation would resume the needs.  I'll do that.

Looking forward to see your tool in action!

**Comment by nikitul** [comment]
Thank you very much.
:)
